src/waveform/renderers/qopengl/shaders/shader.h

```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <QOpenGLShaderProgram>
+
+namespace qopengl {
+class Shader;
+}
+
+class qopengl::Shader : public QOpenGLShaderProgram {
+  public:
+    Shader();
+    ~Shader();
+
+  protected:
+    void load(const QString& fragmentShader, const QString& vertexShader);
```

Comment:
it's not polymorphic (no virtuals). all derived shaders use the load function of the base class, that's all.

Reply:
shouldn't we be using modern c++ delete instead of this macro? it's a clang-tidy modernization:

https://clang.llvm.org/extra/clang-tidy/checks/modernize/replace-disallow-copy-and-assign-macro.html

Reply:
Yes, explicitly doing using `= delete` is better. The `DISALLOW_COPY_AND_ASSIGN` does exactly that anyways.

If you're not using polymorpism anyways, then I see little reason to use inheritance here as well. I personally consider using inheritance for the sole reason of code reuse to be generally an antipattern. I assume you disagree?

Reply:
I extend QOpenGLShaderProgram with a convenience function, ```load(..)```. If instead I inherit all concrete shader classes from QOpenGLShaderProgram directly, I would either have to duplicate this ```load(..)``` function in all of them, or I would need to make it a global function, where I pass the shader's this as argument. Which might work in this case, if it only uses public members of QOpenGLShaderProgram, but who knows. And the first thing I think when I see a function with this as argument is: why isn't this a method :-)

But maybe I am missing something!